pre_commit/languages/rust.py

```diff
@@ -52,15 +93,57 @@ def _add_dependencies(
         f.truncate()
 
 
+def install_rust_with_toolchain(toolchain: str) -> None:
+    # Check if rustup is already present in PATH
+    rustup_exe = parse_shebang.find_executable('rustup')
+    with tempfile.TemporaryDirectory() as rustup_dir:
+        rustup_env = (
+            ('RUSTUP_HOME', rustup_dir),
+        )
+        with envcontext(rustup_env):
```

Comment:
I *think* you can do this using `ExitStack` since creating this directory and such is conditional? (that way it doesn't need to write a directory and change the environment for noops)